src/ofono.c

```diff
@@ -632,15 +732,36 @@ static void ofono_signal_card_removed(GDBusConnection *conn, const char *sender,
 	const char *card = NULL;
 	g_variant_get(params, "(&o)", &card);
 
-	struct ba_transport *t = NULL;
-	if ((t = ofono_transport_lookup(card)) == NULL) {
-		error("Couldn't lookup transport: %s: %s", card, strerror(errno));
+	struct ofono_card_data *ocd;
+
+	if ((ocd = g_hash_table_lookup(ofono_card_data_map, card)) == NULL) {
+		error("Couldn't lookup oFono card data: %s", card);
 		return;
 	}
 
+	struct ba_adapter *a = NULL;
+	struct ba_device *d = NULL;
+	struct ba_transport *t = NULL;
+
+	if ((a = ba_adapter_lookup(ocd->hci_dev_id)) == NULL)
+		goto fail;
+	if ((d = ba_device_lookup(a, &ocd->bt_addr)) == NULL)
+		goto fail;
+	if ((t = ba_transport_lookup(d, ocd->modem_path)) == NULL)
+		goto fail;
+
 	debug("Removing oFono card: %s", card);
 	ba_transport_destroy(t);
 
+fail:
+	if (a != NULL)
+		ba_adapter_unref(a);
+	if (d != NULL)
+		ba_device_unref(d);
+	if (t != NULL)
+		ba_transport_unref(t);
+
+	g_hash_table_remove(ofono_card_data_map, ocd);
```

Comment:
I'm not sure whether this is OK. The `g_hash_table_remove` removes element by the key, but the `ocd` is not a key but a value. Instead of duplicating `ofono_transport_lookup` logic here, I would simply call `g_hash_table_remove(ofono_card_data_map, card);`. As I see it now, the ` ofono_remove_all_cards` can also free resources by calling `g_hash_table_remove_all`.